feat(health): fold successful healthchecks

ateoi opened the pull request:
Fold successful health checks. This feature can be disable by setting the `g:checkhealth_folding_enable` to `v:false`.

Resolves #22796

ateoi committed:
feat(health): fold successful healthchecks

clason commented:
Thank you!

I'd prefer not to introduce a new config variable; just respect the global `fold` options -- either you like folds and want ALL THE THINGS folded, or you don't (and then nothing should be folded).

ateoi committed:
feat(health): healthcheck folding

If `foldenable`:

- fold all sections (failed or not)
- foldtext of checks with warning or error will be prefixed by 'WE'

ateoi committed:
Remove debug message

ateoi committed:
Remove debug message

ateoi committed:
Merge branch 'issue-22796' of github.com:ateoi/neovim into issue-22796

ateoi committed:
Fix unit test

ateoi commented:
> 
> 1. fold _all_ sections (failed or not)
> 2. indicate the status (fail, success, warning) via `foldtext`.
> 

Implemented the suggested behavior. If `foldenable`, all sections will be folded, and the checks with warning or error will be prefixed by `WE`.

![checkhealth](https://user-images.githubusercontent.com/8385547/229951912-e546f34d-3e28-421c-b425-37d3c52595fc.png)

dundargoc commented:
>  either you like folds and want ALL THE THINGS folded, or you don't (and then nothing should be folded).

I disagree with this assessment. I don't want to block this PR, as I believe it's an improvement over the current situation, so I'm only saying this for future discussions. We can merge this as is.

The reason folds are controversial IMO is that what information should be hidden in a normal codebase is highly subjective. Simply put: we can't say with any certainty which parts of a codebase holds important information or not, so we can't be entirely sure whether we can safely hide code in folds or not. This is not the case with healthchecks: either the user has passed or failed the tests the plugin maintainer has deemed important.

If this is more of an objection to folds as how they work in practice, rather than a philosophical objection ("information should NEVER be hidden" kind of thing), then that is another matter entirely, and we can iterate on how to improve that (perhaps keymaps for healtcheck buffers, with keypresses determining the log level of the checks shown?).

clason commented:
Hmm, I see your point, but don't fully agree with it. Let me break it down a bit:

1. Folds are controversial since "hidden text" is controversial (same as conceal; if I wanted invisible formatting characters, I'd be using Word). So _if_ there's folding (which I'm not against), it _must_ respect the global setting -- this is a red line for me.
2. Adding configuration options is not free, and I strongly believe they are not worth it here. Whether failed sections are folded or not is something I personally do not care about (since I'll never see it either way, see 1.) But as the point here is to maximize the amount of information at a glance, I do believe folding everything serves that purpose better. But maybe this can be tuned with different fold levels? There's also the question about how to treat warnings but not errors (or hints). 

Also, I think the case for folding becomes stronger once we have https://github.com/neovim/neovim/pull/20750, as then we can color-code the headings so you can see at a glance "all green" or "some red/yellow".

dundargoc commented:
> So if there's folding (which I'm not against), it must respect the global setting -- this is a red line for me.

Yes, you're right. I was imagining folding as a mechanism for hiding information in this case rather than folding code, but that's a copout in hindsight. Folds are folds and must respect fold options.

> But as the point here is to maximize the amount of information at a glance, I do believe folding everything serves that purpose better.

Small nit: I'd say the point is to filter information rather than maximizing it, but I believe we essentially mean the same thing.

> But maybe this can be tuned with different fold levels?

I think this'd make sense. We already have a syntax files for the checkhealth buffer/filetype, perhaps these can be augmented with "sane" default folding behaviors (sane being my proposal naturally)?

> Also, I think the case for folding becomes stronger once we have https://github.com/neovim/neovim/pull/20750, as then we can color-code the headings so you can see at a glance "all green" or "some red/yellow".

Oooh, that looks nice.

In any case, I don't want to move the goalpost from this PR too much. I think this is perfectly fine for now and something we can iterate on in future PRs.

justinmk reviewed (approved):
Let's try this out, to unblock https://github.com/neovim/neovim/pull/23561

- resolve merge conflict
- add news.txt item
- add a small test case, or just update the existing test instead of setting `nofoldenable`

justinmk commented on test/functional/plugin/health_spec.lua:
```diff
@@ -227,7 +227,7 @@ describe('health.vim', function()
       })
       command("checkhealth foo success1")
-      command("set nowrap laststatus=0")
+      command("set nofoldenable nowrap laststatus=0")
```
should probably have a test case for folding

ateoi committed:
feat: add fold test

ateoi committed:
Merge branch 'master' into issue-22796

ateoi committed:
fix: update fold test

ateoi committed:
feat: update news.txt

ateoi committed:
revert: "feat: update news.txt"

This reverts commit 4571407557b2ada315935c0f65ead77fe069c0e7.

ateoi commented:
> * add news.txt item

How can I update `news.txt` without breaking the tests?

dundargoc commented:
You need to change `|foldenable|` to `|'foldenable'|`, since there is no tag without the ''.

justinmk commented on runtime/lua/vim/health.lua:
```diff
@@ -2,6 +2,40 @@ local M = {}
+  if vim.b.failedchecks[foldtext] == nil then
+    local warning = '- WARNING '
+    local warninglen = string.len(warning)
+    local error = '- ERROR '
+    local errorlen = string.len(error)
```
```suggestion
    local err = '- ERROR '
    local errlen = string.len(err)
```

ateoi committed:
feat: update news.txt

justinmk commented on runtime/doc/news.txt:
```diff
@@ -104,6 +104,9 @@ The following changes to existing APIs or features add new behavior.
   as Lua code regardless of its extension.
 
+• |:checkhealth| buffer now implements |folding|. The initial folding status is
+  defined by the |'foldenable'| option.
```
```suggestion
  defined by the 'foldenable' option.
```

justinmk committed:
Update runtime/doc/news.txt

ateoi committed:
fix: improve checkhealth buffer detection

ateoi committed:
refactor: don't shadow a builtin

clason commented:
Hmm, would have preferred this to be opt-in (I forgot that `foldenable` is set by default.)

clason commented:
No no, foldenable is going to be set to false, with a sledge hammer.

(I _hate_ folds with a passion.)

justinmk commented:
`zR` expands all folds. For https://github.com/neovim/neovim/pull/23561 ,  "opt-out" is kind of required?

clason commented:
So does `zi`; both add an extra step for using `checkhealth` as general diagnostics (which nvim-treesitter has started to lean into). `vim.o.foldenable = false` it is.
